x/slashing/keeper/migrations.go

```diff
@@ -118,7 +123,92 @@ func (m Migrator) Migrate2to3(ctx sdk.Context) error {
+		store.Set(types.ValidatorMissedBlockBitArrayKey(consAddrKey), bz)
+	}
+	ctx.Logger().Info("Done migrating")
+	return nil
+}
+
+// Migrate3to4 migrates from version 3 to 4.
+func (m Migrator) Migrate3to4(ctx sdk.Context) error {
+	ctx.Logger().Info("Migrating 3 -> 4")
+	store := ctx.KVStore(m.keeper.storeKey)
+	valMissedMap := make(map[string]types.ValidatorMissedBlockArray)
+	ctx.Logger().Info("Migrating Signing Info")
+	signInfoIter := sdk.KVStorePrefixIterator(store, types.ValidatorSigningInfoKeyPrefix)
+	newSignInfoKeys := [][]byte{}
+	newSignInfoVals := []types.ValidatorSigningInfo{}
+	// use previous height to calculate index offset
+	window := m.keeper.SignedBlocksWindow(ctx)
+	index := window - 1
+	defer signInfoIter.Close()
+	for ; signInfoIter.Valid(); signInfoIter.Next() {
+		ctx.Logger().Info(fmt.Sprintf("Migrating Signing Info for key: %v\n", signInfoIter.Key()))
+		var oldInfo types.ValidatorSigningInfoLegacyMissedHeights
+		m.keeper.cdc.MustUnmarshal(signInfoIter.Value(), &oldInfo)
+
+		newInfo := types.ValidatorSigningInfo{
+			Address:             oldInfo.Address,
+			StartHeight:         oldInfo.StartHeight,
+			IndexOffset:         index,
+			JailedUntil:         oldInfo.JailedUntil,
+			Tombstoned:          oldInfo.Tombstoned,
+			MissedBlocksCounter: oldInfo.MissedBlocksCounter,
+		}
+		newSignInfoKeys = append(newSignInfoKeys, signInfoIter.Key())
+		newSignInfoVals = append(newSignInfoVals, newInfo)
+	}
+	signInfoIter.Close()
+
+	if len(newSignInfoKeys) != len(newSignInfoVals) {
+		return fmt.Errorf("new sign info data length doesn't match up")
+	}
+	ctx.Logger().Info("Writing New Signing Info")
+	for i := range newSignInfoKeys {
+		bz := m.keeper.cdc.MustMarshal(&newSignInfoVals[i])
+		store.Set(newSignInfoKeys[i], bz)
+	}
+
+	// need to turn this into a bool array
+	ctx.Logger().Info("Migrating Missed Block Bit Array")
+	startWindowHeight := ctx.BlockHeight() - window
+	iter := sdk.KVStorePrefixIterator(store, types.ValidatorMissedBlockBitArrayKeyPrefix)
+	defer iter.Close()
+	for ; iter.Valid(); iter.Next() {
+		var missedInfo types.ValidatorMissedBlockArrayLegacyMissedHeights
+		key := iter.Key()
+		consAddrBytes := key[2:]
+
+		consAddr := sdk.ConsAddress(consAddrBytes)
+		ctx.Logger().Info(fmt.Sprintf("Migrating for next validator with consAddr: %s\n", consAddr.String()))
+
+		newBoolArray := make([]bool, window)
+		m.keeper.cdc.MustUnmarshal(iter.Value(), &missedInfo)
+		heights := missedInfo.MissedHeights
+		for _, height := range heights {
+			if height < startWindowHeight {
+				continue
+			}
+			index := height - startWindowHeight
+			newBoolArray[index] = true
+		}
+
+		valMissedMap[consAddr.String()] = types.ValidatorMissedBlockArray{
+			Address:      missedInfo.Address,
+			MissedBlocks: m.keeper.ParseBoolArrayToBitGroups(newBoolArray),
+			WindowSize:   window,
+		}
+	}
+
+	ctx.Logger().Info("Writing new validator missed blocks infos")
+	for key, missedBlockArray := range valMissedMap {
```

Comment:
this should be fine, but we should sort this anyways to avoid any source of nondeterm